### PR TITLE
Feature/logs

### PR DESCRIPTION
**Pull request by croomes**
*Feature requires StorageOS 0.10 (not yet released)*

Adds logs streaming with `storageos logs -f`.  The current PR does a single node only (where STORAGEOS_HOST points to), but we expect to add support to consolidate logs from all cluster nodes.

### CODE REVIEW

**Comment by croomes** [comment]
Log consolidation was added at the server end

File: cli/command/formatter/log_stream.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+			entry.Level = level
+		case "msg":
+			entry.Message = str(v)
+		default:
+			entry.Data[k] = v
+		}
+
+	}
+	return entry, nil
+}
```
Comment (darkowlzz) [inline cli/command/formatter/log_stream.go]: Can we add tests for this marshaling, just to ensure it works as expected? :)

File: cli/command/formatter/logrus.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+			f.appendKeyValue(b, "msg", entry.Message)
+		}
+		for _, key := range keys {
+			f.appendKeyValue(b, key, entry.Data[key])
+		}
+	}
+
+	b.WriteByte('\n')
+	return b.Bytes(), nil
+}
```
Comment (darkowlzz) [inline cli/command/formatter/logrus.go]: Too many conditionals here. Some tests would be great.

File: cli/command/formatter/logrus.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+		fmt.Fprintf(b, "\x1b[%dm%s\x1b[0m[%04d] %-44s ", levelColor, levelText, int(entry.Time.Sub(baseTimestamp)/time.Second), entry.Message)
+	} else {
+		fmt.Fprintf(b, "\x1b[%dm%s\x1b[0m[%s] %-44s ", levelColor, levelText, entry.Time.Format(timestampFormat), entry.Message)
+	}
+	for _, k := range keys {
+		v := entry.Data[k]
+		fmt.Fprintf(b, " \x1b[%dm%s\x1b[0m=", levelColor, k)
+		f.appendValue(b, v)
+	}
+}
```
Comment (darkowlzz) [inline cli/command/formatter/logrus.go]: Too many conditionals here. Some tests would be great.

File: cli/command/logs/follow.go [hunk 4]
```diff
@@ -1,9 +1,107 @@
...
+
+	c.SetPongHandler(func(string) error { return c.SetReadDeadline(time.Now().Add(wait)) })
+	c.SetReadLimit(maxMessageSize)
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	done := make(chan struct{})
+
+	// Incoming message handler
```
Comment (darkowlzz) [inline cli/command/logs/follow.go]: This is fine but keeping the run functions small helps in isolating the main code and adding test to validate the correctness of the code, wherever possible. How about moving this handlers out as separate function?

File: cli/command/websocket.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	for _, h := range cli.GetHosts() {
+		u, err := url.Parse(h)
+		if err != nil {
+			continue
+		}
+		u.Scheme = "ws"
+		urls = append(urls, u)
+	}
+	return urls
+}
```
Comment (darkowlzz) [inline cli/command/websocket.go]: Some tests would be great.

**Comment by croomes** [comment]
Agree with the comments about unit tests, opened a ticket and we will add across all commands.

**Pull request merged**